### PR TITLE
[Feature] Support Array<String>work on low-cardinality

### PR DESCRIPTION
**Pull request by Seaven**
Why I'm doing:

What I'm doing:
base on https://github.com/StarRocks/starrocks/pull/34246

support array<String> on low-cardinality
* support array function: array_min/array_distinct/array[x]...
    * array column doesn't need decodeExpr to define a new dict column, do expression directly (use origin dict-id)
* update DictMappingExpr, support execute array function first and do decode later
* update DictMappingExpr's explain, add `dictDefine` and `dictDecode`
* storage only support global dict optimize, don't support local dict optimize
* forbidden push down array-predicate to storage
* refactor array_distinct, promise order now

```
+-------------------------------------------------------------------------------------------------+
| Explain String                                                                                  |
+-------------------------------------------------------------------------------------------------+
| PLAN FRAGMENT 0                                                                                 |
|  OUTPUT EXPRS:5: array_max | 10: array_min                                                      |
|   PARTITION: UNPARTITIONED                                                                      |
|                                                                                                 |
|   RESULT SINK                                                                                   |
|                                                                                                 |
|   7:MERGING-EXCHANGE                                                                            |
|      limit: 2                                                                                   |
|                                                                                                 |
| PLAN FRAGMENT 1                                                                                 |
|  OUTPUT EXPRS:                                                                                  |
|   PARTITION: RANDOM                                                                             |
|                                                                                                 |
|   STREAM DATA SINK                                                                              |
|     EXCHANGE ID: 07                                                                             |
|     UNPARTITIONED                                                                               |
|                                                                                                 |
|   6:TOP-N                                                                                       |
|   |  order by: <slot 5> 5: array_max ASC                                                        |
|   |  offset: 0                                                                                  |
|   |  limit: 2                                                                                   |
|   |                                                                                             |
|   5:HASH JOIN                                                                                   |
|   |  join op: INNER JOIN (BROADCAST)                                                            |
|   |  colocate: false, reason:                                                                   |
|   |  equal join conjunct: 10: array_min = 5: array_max                                          |
|   |                                                                                             |
|   |----4:EXCHANGE                                                                               |
|   |                                                                                             |
|   1:Project                                                                                     |
|   |  <slot 10> : DictDecode(11: a1, [<place-holder>], array_min(11: a1))                        |
|   |                                                                                             |
|   0:OlapScanNode                                                                                |
|      TABLE: s2                                                                                  |
|      PREAGGREGATION: ON                                                                         |
|      PREDICATES: DictDecode(11: a1, [<place-holder> IS NOT NULL], array_min(11: a1))            |
|      partitions=1/1                                                                             |
|      rollup: s2                                                                                 |
|      tabletRatio=10/10                                                                          |
|      tabletList=7455022,7455024,7455026,7455028,7455030,7455032,7455034,7455036,7455038,7455040 |
|      cardinality=6400                                                                           |
|      avgRowSize=17.0                                                                            |
|                                                                                                 |
| PLAN FRAGMENT 2                                                                                 |
|  OUTPUT EXPRS:                                                                                  |
|   PARTITION: RANDOM                                                                             |
|                                                                                                 |
|   STREAM DATA SINK                                                                              |
|     EXCHANGE ID: 04                                                                             |
|     UNPARTITIONED                                                                               |
|                                                                                                 |
|   3:Project                                                                                     |
|   |  <slot 5> : DictDecode(12: a1, [<place-holder>], array_max(12: a1))                         |
|   |                                                                                             |
|   2:OlapScanNode                                                                                |
|      TABLE: s1                                                                                  |
|      PREAGGREGATION: ON                                                                         |
|      PREDICATES: DictDecode(12: a1, [<place-holder> IS NOT NULL], array_max(12: a1))            |
|      partitions=1/1                                                                             |
|      rollup: s1                                                                                 |
|      tabletRatio=10/10                                                                          |
|      tabletList=7455045,7455047,7455049,7455051,7455053,7455055,7455057,7455059,7455061,7455063 |
|      cardinality=49                                                                             |
|      avgRowSize=17.0                                                                            |
+-------------------------------------------------------------------------------------------------+
```

### Preformance
```
MySQL ssb> select count(1) from arr2;
+-----------+
| count(1)  |
+-----------+
| 128000000 |
+-----------+
1 row in set
Time: 0.534s
MySQL ssb> desc arr2;
+-------------+--------------------+------+-------+---------+-------+
| Field       | Type               | Null | Key   | Default | Extra |
+-------------+--------------------+------+-------+---------+-------+
| lo_custkey  | int                | NO   | true  | <null>  |       |
| lo_quantity | array<varchar(50)> | NO   | false | <null>  |       |
| lo_discount | array<varchar(50)> | NO   | false | <null>  |       |
| lo_shipmode | array<varchar(50)> | NO   | false | <null>  |       |
+-------------+--------------------+------+-------+---------+-------+
4 rows in set
Time: 0.006s
MySQL ssb>
```

and array's avg-length is 3, and string's avg-length is 5 bytes, and pipeline_dop = 1

SQLs:
```
-- count distinct
select count(distinct lo_quantity[2]) from arr2; 
-- predicate
select count(1) from arr2 where array_max(lo_discount) = '9b' or array_min(lo_quantity) = '16a';
-- project
select MIN(upper(lo_discount[2])), MAX(lower(lo_shipmode[3])) from arr2 ;
```

| SQL |  Cost(low_cardinality_optimize_v2=false) |  Cost(low_cardinality_optimize_v2=true) |
| -- | -- | -- |
| SQL-1 (count distinct) | 2.762s | 1.947s |
| SQL-2 (predicate) | 8.868s | 4.383s |
| SQL-3 (project) | 4.739s | 3.316s |

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/lowcardinality/DecodeCollector.java [hunk 1]
```diff
@@ -0,0 +1,655 @@
...
+        return info;
+    }
+
+    @Override
+    public DecodeNodeInfo visitPhysicalDistribution(OptExpression optExpression, DecodeNodeInfo context) {
+        if (context.outputStringColumns.isEmpty()) {
+            return DecodeNodeInfo.EMPTY;
+        }
+        DecodeNodeInfo result = context.createOutputInfo();
+        if (context.parent != null && context.parent.getOp() instanceof PhysicalJoinOperator) {
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/lowcardinality/DecodeCollector.java:353]: What the different when the parent is not a joinOp?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/lowcardinality/DecodeCollector.java [hunk 2]
```diff
@@ -0,0 +1,655 @@
...
+
+    private static boolean supportLowCardinality(Type type) {
+        return type.isVarchar() || (type.isArrayType() && ((ArrayType) type).getItemType().isVarchar());
+    }
+
+    // Check if an expression can be optimized using a dictionary
+    // If the expression only contains a string column, the expression can be optimized using a dictionary
+    private static class DictExpressionCollector extends ScalarOperatorVisitor<ScalarOperator, Void> {
+        private static final ScalarOperator CONSTANTS = ConstantOperator.TRUE;
+        private static final ScalarOperator VARIABLES = ConstantOperator.FALSE;
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/lowcardinality/DecodeCollector.java:481]: VARIABLES is a little confued? how about DICT_COL_HODLER ?

**Comment by kangkaisen** [comment]
```
storage only support global dict optimize, don't support local dict optimize
```
why?

File: be/src/runtime/global_dict/parser.cpp [hunk 3]
```diff
@@ -112,13 +159,48 @@ class DictFuncExpr final : public Expr {
                 return res;
             }
         }
-
-        return nullptr;
     }
 
-    Expr* clone(ObjectPool* pool) const override { return pool->add(new DictFuncExpr(_origin_expr, _dict_opt_ctx)); }
+    ColumnPtr _translate_array(ColumnPtr& array, size_t num_rows) {
```
Comment (kangkaisen) [inline be/src/runtime/global_dict/parser.cpp:164]: what's the translate here? encode or decode?

File: be/src/runtime/global_dict/parser.cpp [hunk 4]
```diff
@@ -141,6 +223,10 @@ class DictFuncExpr final : public Expr {
     // data column ptr
     ColumnPtr _data_column_ptr;
 
+    // mark intput column type
+    bool _is_string = false;
```
Comment (kangkaisen) [inline be/src/runtime/global_dict/parser.cpp:227]: is one enum type better?

File: be/src/storage/rowset/column_decoder.h [hunk 5]
```diff
@@ -54,6 +55,11 @@ class ColumnDecoder {
 
     int16_t* code_convert_data() { return _code_convert_map.has_value() ? _code_convert_map->data() + 1 : nullptr; }
 
+private:
+    Status encode_string_to_global_id(Column* datas, Column* codes);
```
Comment (kangkaisen) [inline be/src/storage/rowset/column_decoder.h:59]:
```suggestion
    Status _encode_string_to_global_id(Column* datas, Column* codes);
```

File: fe/fe-core/src/main/java/com/starrocks/analysis/PlaceHolderExpr.java [hunk 6]
```diff
@@ -50,6 +50,7 @@ protected void analyzeImpl(Analyzer analyzer) throws AnalysisException {
     @Override
     protected String toSqlImpl() {
         return "<place-holder>";
+//        return "<place-holder: " + slotId + ">";
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/analysis/PlaceHolderExpr.java:53]: remove

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/lowcardinality/DecodeContext.java [hunk 7]
```diff
@@ -193,7 +202,118 @@ private ColumnRefOperator createNewDictColumn(ColumnRefOperator column) {
         }
     }
 
-    private class FunctionRewriter extends BaseScalarOperatorShuttle {
+    // define mode: means the result column is dict, DictExpr should return int/array<int> type
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/lowcardinality/DecodeContext.java:205]: encode and decode mode? not define mode.

**Comment by Seaven** [comment]
> ```
> storage only support global dict optimize, don't support local dict optimize
> ```
> 
> why?

storage dict optimize isn't same framework with global dict optimize, it's need hack & rewrite expression-tree, It's very difficult now

File: be/src/runtime/global_dict/parser.cpp (hunk 3 above)
Comment (Seaven) [reply be/src/runtime/global_dict/parser.cpp:164]: decode

Comment (Seaven) [reply be/src/runtime/global_dict/parser.cpp:164]: decode, just for keep same name as before

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/lowcardinality/DecodeContext.java (hunk 7 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/lowcardinality/DecodeContext.java:205]: I want to express is that this column is a new dict-column and it's defined by the expression, it's not encode string column to dict column

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/lowcardinality/DecodeCollector.java [hunk 8]
```diff
@@ -369,12 +388,15 @@ public DecodeInfo visitPhysicalOlapScan(OptExpression optExpression, DecodeInfo
...
             }
 
-            ColumnStatistic columnStatistic =
-                    GlobalStateMgr.getCurrentStatisticStorage().getColumnStatistic(table, column.getName());
+
+            ColumnStatistic columnStatistic = GlobalStateMgr.getCurrentStatisticStorage()
+                    .getColumnStatistic(table, column.getName());
             // Condition 2: the varchar column is low cardinality string column
             if (!FeConstants.USE_MOCK_DICT_MANAGER && (columnStatistic.isUnknown() ||
                     columnStatistic.getDistinctValuesCount() > CacheDictManager.LOW_CARDINALITY_THRESHOLD)) {
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/lowcardinality/DecodeCollector.java:402]: array column's ndv always 1 in statistics now, and I was supported BE's array-dict collect

**Review by packy92 — DISMISSED** [review]

File: be/src/column/field.cpp [hunk 9]
```diff
@@ -44,4 +44,23 @@ ColumnPtr Field::create_column() const {
     return ChunkHelper::column_from_field(*this);
 }
 
+FieldPtr Field::convert_to_dict_field(const Field& field) {
+    if (field.type()->type() == TYPE_VARCHAR) {
+        FieldPtr res = std::make_shared<Field>(field);
+        res->_type = get_type_info(TYPE_INT);
+        return res;
+    } else if (field.type()->type() == TYPE_ARRAY && field.sub_field(0).type()->type() == TYPE_VARCHAR) {
```
Comment (fzhedu) [inline be/src/column/field.cpp:52]: don't support nested string array? like array<array<string>>

Comment (fzhedu) [reply be/src/column/field.cpp:52]: can it support TYPE_BINARY?

Comment (Seaven) [reply be/src/column/field.cpp:52]: don't support now

File: be/src/exprs/array_functions.tpp [hunk 10]
```diff
@@ -101,26 +101,25 @@ private:
...
             }
-        }
-
-        auto& dest_data_column = dest_column->elements_column();
-        auto& dest_offsets = dest_column->offsets_column()->get_data();
 
-        if (has_null) {
-            dest_data_column->append_nulls(1);
+            const auto& tt = item.get<CppType>();
+            if (hash_set->count(tt) == 0) {
```
Comment (fzhedu) [inline be/src/exprs/array_functions.tpp:117]: does count() should iterator all of element in the set? seems not efficient

File: be/src/exprs/dictmapping_expr.cpp [hunk 11]
```diff
@@ -34,17 +34,32 @@ StatusOr<ColumnPtr> DictMappingExpr::evaluate_checked(ExprContext* context, Chun
...
-    auto data_column = ColumnHelper::get_data_column(target_column.get());
 
-    if (data_column->is_binary()) {
-        DCHECK(dict_func_expr == nullptr);
-        return get_child(1)->evaluate_checked(context, ptr);
-    } else if (dict_func_expr != nullptr) {
-        return dict_func_expr->evaluate_checked(context, ptr);
-    } else {
-        return Status::InternalError("unreachable path, dict_func_expr shouldn't be nullptr");
+    if (_children.size() == 2) {
```
Comment (stdpain) [inline be/src/exprs/dictmapping_expr.cpp:38]: Explain when children==2 and when children==3.

File: be/src/runtime/global_dict/decoder.cpp [hunk 12]
```diff
@@ -16,52 +16,108 @@
...
+        auto* const_column = down_cast<ConstColumn*>(in);
+        auto* in_array = down_cast<ArrayColumn*>(const_column->data_column().get());
+        auto in_element = in_array->elements_column();
+        auto in_offsets = in_array->offsets_column();
+
+        auto* out_array = down_cast<ArrayColumn*>(out);
+        auto out_element = out_array->elements_column();
+        auto out_offsets = out_array->offsets_column();
+        out_offsets->swap_column(*in_offsets);
+        return decode_string(in_element.get(), out_element.get());
```
Comment (stdpain) [inline be/src/runtime/global_dict/decoder.cpp:67]: For nullable arrays, a slice in a partial null value is an illegal value, and I'm not sure they're safe.

Comment (stdpain) [reply be/src/runtime/global_dict/decoder.cpp:67]: logical value:
[null, "key2", "key3"]
physical value:
nullcolumn: [1, 0, 0]
arraycolumn: ["key1", "key2", "key3"]

File: be/src/storage/rowset/dictcode_column_iterator.cpp [hunk 13]
```diff
@@ -94,9 +127,14 @@ Status GlobalDictCodeColumnIterator::build_code_convert_map(ScalarColumnIterator
     return Status::OK();
 }
 
-ColumnPtr GlobalDictCodeColumnIterator::_new_local_dict_col(bool nullable) {
+ColumnPtr GlobalDictCodeColumnIterator::_new_local_dict_col(Column* src) {
     ColumnPtr res = std::make_unique<Int32Column>();
-    if (nullable) {
+    auto code_data = ColumnHelper::get_data_column(src);
+    if (code_data->is_array()) {
+        res = ArrayColumn::create(NullableColumn::create(res, NullColumn::create()), UInt32Column::create());
```
Comment (stdpain) [inline be/src/storage/rowset/dictcode_column_iterator.cpp:134]: Is this an implicit constraint?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/DictMappingOperator.java [hunk 14]
```diff
@@ -24,16 +24,27 @@
 import java.util.Objects;
 
 public class DictMappingOperator extends ScalarOperator {
-
+    // use dict id
     private ColumnRefOperator dictColumn;
+    // dict expression
     private ScalarOperator originScalaOperator;
+    // input string expression
+    private ScalarOperator stringScalarOperator;
```
Comment (stdpain) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/DictMappingOperator.java:32]: rename to StringColumnProviderOperator

File: be/src/exprs/dictmapping_expr.cpp (hunk 11 above)
Comment (Seaven) [reply be/src/exprs/dictmapping_expr.cpp:38]: children == 2: DictExpr(string column, string expression)
children == 3: DictExpr(array column, string expression, array expression)
do array-expresion first, then string expression

File: be/src/runtime/global_dict/decoder.cpp (hunk 12 above)
Comment (Seaven) [reply be/src/runtime/global_dict/decoder.cpp:67]: don't worry, if string column has any value in actual, it's should has dict-id?

File: be/src/storage/rowset/dictcode_column_iterator.cpp (hunk 13 above)
Comment (Seaven) [reply be/src/storage/rowset/dictcode_column_iterator.cpp:134]: I think not, dict-iterator read interface must keep same as normal-iterator interface

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/DictMappingOperator.java (hunk 14 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/DictMappingOperator.java:32]: good idea

File: be/src/exprs/array_functions.tpp (hunk 10 above)
Comment (Seaven) [reply be/src/exprs/array_functions.tpp:117]: i think it's not key-point in th function, and should be better than before

File: be/src/storage/rowset/dictcode_column_iterator.cpp [hunk 15]
```diff
@@ -47,8 +75,13 @@ Status GlobalDictCodeColumnIterator::decode_dict_codes(const Column& codes, Colu
 
     if (output_nullable) {
         // reserve null data
-        down_cast<NullableColumn*>(words)->null_column_data().resize(size);
+        auto word_nulls = down_cast<NullableColumn*>(words)->mutable_null_column();
+        down_cast<NullableColumn*>(words)->set_has_null(codes.has_null());
         const auto& null_data = down_cast<const NullableColumn&>(codes).immutable_null_column_data();
+        word_nulls->resize(0);
+        for (size_t i = 0; i < size; ++i) {
+            word_nulls->append(null_data[i]);
```
Comment (stdpain) [inline be/src/storage/rowset/dictcode_column_iterator.cpp:83]: prefer memcpy instead of vector::append

**Review by stdpain — DISMISSED** [review]

File: test/sql/test_low_cardinality/R/test_delete_conditions [hunk 16]
```diff
@@ -15,6 +15,7 @@ insert into t0 values (null, null, null, null);
 -- !result
 [UC] analyze full table t0;
 -- result:
+test_db_d0bb90d29d5011eeaa6400163e04d4c2.t0	analyze	status	OK
```
Comment (stdpain) [inline test/sql/test_low_cardinality/R/test_delete_conditions:18]: remove it

**Review by packy92 — DISMISSED** [review]

File: be/src/storage/rowset/dictcode_column_iterator.cpp (hunk 15 above)
Comment (Seaven) [reply be/src/storage/rowset/dictcode_column_iterator.cpp:83]: update later

**Review by packy92 — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**